Walk compressed IL -> Native map for EventTrace and stacktrace symbolication scenarios

davidwrighton opened the pull request:
Add paths for reading the IL-Native map which doesn't create a DebuggerFunctionInfo or DebuggerJitInfo
Use it in the stackwalker code for EH, and for the event tracing MethodToILMap

This process is somewhat slower than the path through the diagnostic infrastructure, but it has no expensive locks required. To mitigate the performance loss, the format of the compressed bounds information was changed to a an encoding where all entries in the table are fixed size, but the actual bit widths of the various fields are optimally small. Overall this results in a size savings of about 50KB on System.Private.CoreLib.dll, as well as capturing back all of the performance costs of this change for relatively small functions.

In combination with the NativeToIL map cache added a few days ago, this allows us to eliminate locks from the native offset to il offset path, which should substantially reduce the cost when heavily multithreaded applications are experiencing significant error rates. (And the cache added mitigates the remaining performance loss from using the compressed data instead of the uncompressed form.)

In addition, by changing the event trace path to use the same approach, we will remove the `DebuggerJitInfo` generation path from most customers non-debugging scenarios. (It still remains in the ICorProfiler:GetILToNativeMapping apis, but usage of that api is relatively rare).

There is validation which utilizes the debugger apis to validate that the behavior hasn't changed at all.

davidwrighton committed:
First pass at removing use of the debugger copy of the IL to native map
- This removes it from normal path
- This also adds an assert that it is correct

Logically, this code should probably replace the current implementation of GetILOffsetFromNative but this first pass is just to get the code working

This is untested, but it does compile.

davidwrighton committed:
Get the helper to work correctly

davidwrighton committed:
More fixes

davidwrighton committed:
Remove accidental character added to PR

davidwrighton committed:
- Make the NibbleReader a bit faster by loading data in pointer sized chunks and doing very cheap operations per nibble read

- Add a fixed size cache around the native to IL mapping operations in the stacktrace logic
  - Unfortunately, simply reading the nibble data is pretty slow, so this is needed to avoid regressing performance
  - With this change the performance of newing up a StackTrace object on a multithreaded environment is primarily driven by GC allocation rate, and not significantly by any other locks.

davidwrighton committed:
Missed change to COOPERATIVE mode in contract

davidwrighton committed:
Fix epilog handling

davidwrighton committed:
Add assertions around correctness of offset data

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into UseWalkILOffsetInsteadOfDebuggerData

davidwrighton committed:
Fix IL offset walk for epilog discovery

davidwrighton committed:
Get valid result details more correct.

davidwrighton committed:
CALL_INSTRUCTION mapping data affects the greatest IL offset found calculation

davidwrighton committed:
Faster nibble reading in checked builds

davidwrighton committed:
Faster underlying compressed format
This both generates less data (Experiments show about a 70KB reduction in size of System.Private.CoreLib.dll R2R image)
And the reader as implemented here, is quite a bit faster

When combined with the cache constructed in a different part of this PR, this results in the cost of IL offset determination for a given IP to have negligible cost.

davidwrighton committed:
Fix Unix build breaks

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -13215,6 +13216,9 @@ static TADDR UnsafeJitFunctionWorker(
         {
             g_pDebugInterface->JITComplete(nativeCodeVersion, (TADDR)nativeEntry);
+#ifdef DEBUG
+            ValidateILOffsets(ftn, NULL, 0, (uint8_t*)nativeEntry, sizeOfCode);
```
This offset validation is causing some tests to time out. Build analysis may be misclassifying some of these timeouts against unrelated known issues (e.g. https://github.com/dotnet/runtime/issues/116748)

davidwrighton replied:
Agreed. I don't think I can keep this in the PR when we check in(or rather, the Validation will need to move to a less likely spot, but its been quite helpful for identifying issues.

davidwrighton replied:
Instead of moving the check, I've simply adjusted the step through the method so that the total work is O(N) instead of O(N^2).  That should mitigate this problem at the expense of the check being less perfect.

davidwrighton committed:
Fix Unix build break

davidwrighton committed:
Fix assertion failures due to missing case in VarLoc comparison routine

davidwrighton committed:
Enable R2R to have the new compressed format

davidwrighton committed:
Update ILOffsets format for R2R at the reader side

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into UseWalkILOffsetInsteadOfDebuggerData

davidwrighton committed:
Fix merge issue

davidwrighton committed:
Fix errors noted in testing

davidwrighton committed:
- Fix interpreter scenarios
- Add handling for event tracing uses of IL to Native mapping logic to avoid using the diagnostic infrastructure there
- Add validation logic for the event tracing path the ValidateILOffsets
- Limit the NativeOffset to IL Offset validation logic to do O(N) work instead of O(N*N) work in worst case.

davidwrighton committed:
Handle Instrumented IL mappings

davidwrighton committed:
Fix some more issues

davidwrighton committed:
Self-feedback on the PR

davidwrighton committed:
Reduce down to only 1 implementation of bounds processing through using a template instead of 2 seperate implementations

davidwrighton committed:
Fix comparison failures found in GCC build leg

davidwrighton committed:
Fix wasm build

davidwrighton committed:
Fix comparison failures by always processing all the data even though we'll only use a portion of it

noahfalk commented on src/coreclr/vm/debugdebugger.cpp:
```diff
@@ -1254,4 +1607,192 @@ void DebugStackTrace::Element::InitPass2()
+    return i + 1;
+}
+
+void ILToNativeMapArrays::QuickSort(uint32_t* rguiNativeOffset, uint32_t *rguiILOffset, int32_t low, int32_t high, int32_t stackDepth)
```
I see that utilcode has a CQuickSort implementation. Should we be leveraging that or there is some reason it was unsuitable?

noahfalk commented on src/coreclr/vm/debuginfostore.cpp:
```diff
@@ -310,6 +357,61 @@ static void DoBounds(
+
+    // Bounds info contains (Native Offset, IL Offset, flags)
+    // - Sorted by native offset (so use a delta encoding for that).
+    // - IL offsets aren't sorted, but they should be close to each other (so a signed delta encoding)
```
It didn't look like we were using signed delta encoding for IL offsets in the implementation.

noahfalk reviewed (commented):
Thanks David! This looks very nice overall. A few comments inline. I think there may be some unintended changed behavior in GetILToNativeMapping() that would cause trouble for profilers.

noahfalk commented:
fyi @dotnet/dotnet-diag

davidwrighton commented on src/coreclr/debug/ee/debugger.cpp:
```diff
@@ -2879,7 +2882,7 @@ HRESULT Debugger::GetILToNativeMapping(PCODE pNativeCodeStartAddress, ULONG32 cM
             if (cMap == 0)
             {
-                if (DebugInfoManager::GetBoundariesAndVars(diq, nullptr, nullptr, pcMap, nullptr, nullptr, nullptr))
+                if (DebugInfoManager::GetBoundariesAndVars(diq, nullptr, nullptr, true /* preferInstrumentedBounds */, pcMap, nullptr, nullptr, nullptr))
```
Ah, yeah, this IS a behavior change. Looking at the behavior, its probably a desirable behavior change, since the `SequenceMap` stored in the `DebuggerJitInfo` represents the uninstrumented IL offsets, but given that I'm trying so hard to avoid any visible behavior changes, I'll swap it around.

davidwrighton commented on src/coreclr/vm/debugdebugger.cpp:
```diff
@@ -1180,6 +1180,365 @@ void InsertIntoNativeToILCache(void* ip, bool fAdjustOffset, uint32_t dwILOffset
+        pWalkData->greatestILOffsetFound = (int)pOffsetMapping->ilOffset;
+    }
+
+    // Also, ignore all the callsite mappings since they are not relevant for the standard Native-IL mapping
```
Done

davidwrighton commented on src/coreclr/debug/ee/functioninfo.cpp:
```diff
@@ -1045,6 +1046,12 @@ void DebuggerJitInfo::SetBoundaries(ULONG32 cMap, ICorDebugInfo::OffsetMapping *
+    //  since we now have the ability to get an instrumented IL offset mapping
+    // directly from the VM. This work was not done when adding the instrumented
+    // IL mapping due to the work ocurring too close to the shipping deadline for .NET 10.
+    // If we do so, we need to change the input to this function to be the instrumented IL offset mapping
```
Fixed

davidwrighton committed:
Feedback

davidwrighton committed:
Merge branch 'UseWalkILOffsetInsteadOfDebuggerData' of https://github.com/davidwrighton/runtime into UseWalkILOffsetInsteadOfDebuggerData

davidwrighton commented on src/coreclr/vm/debuginfostore.cpp:
```diff
@@ -623,6 +957,23 @@ PTR_BYTE CompressDebugInfo::CompressBoundariesAndVars(
         pBounds = boundsBuffer.GetBlob(&cbBounds);
     }
 
+    NibbleWriter instrumentedILBoundsBuffer;
```
I've used an enum class to make this much clearer and gotten rid of the bool. Now its much more obvious, and I fixed the naming reversal.

noahfalk reviewed (approved):
👍

davidwrighton committed:
Merge branch 'main' into UseWalkILOffsetInsteadOfDebuggerData

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into UseWalkILOffsetInsteadOfDebuggerData

davidwrighton committed:
Fixup changes from merge
